Custom serialization

jasongrout opened the pull request:
This is an attempt to address #7729.

This depends on #7798, #7801, and #7780, which have all been merged.

jasongrout commented:
@jdfreder, @sylvaincorlay - this isn't nearly ready yet, but it's getting there.  Just FYI

jdfreder commented:
I'll wait to try to wrap my head around the code entirely until it's in a working state, just ping when.

jdfreder commented on IPython/html/static/widgets/js/widget.js:
```diff
@@ -136,12 +136,48 @@ define(["widgets/js/manager",
+                                    })(state[k]));
+                                }
+                            }
+                            return utils.resolve_promises_dict(state);
```
The block above should get moved into it's own method.

jasongrout replied:
I agree; that's a bit of a beast.

jasongrout replied:
I rewrote it and think it's a lot easier to follow now.

jdfreder commented on IPython/html/static/widgets/js/widget.js:
```diff
@@ -164,7 +200,7 @@ define(["widgets/js/manager",
             var that = this;
             // Handle when a widget is updated via the python side.
-            return this._unpack_models(state).then(function(state) {
+            return Promise.resolve(state).then(function(state) {
```
Isn't the resolve here redundant?  Above you use `return utils.resolve_promises_dict(state);` to resolve the state dict before a `then` which gets the resolved contents.

jasongrout replied:
I created one more promise than necessary.  I could have done
```javascript
return new Promise(function(resolve, reject) {
    ....
    resolve();
})
```
but thought it would be simpler to just resolve the state to create a promise and chain a `then`.  Do you have a preference?

jasongrout replied:
I just changed it to use `new Promise` - you can see if you like that better.

jdfreder commented on IPython/html/static/widgets/js/widget.js:
```diff
@@ -136,12 +136,48 @@ define(["widgets/js/manager",
+                                if (metadata[k] && metadata[k].serialization) {
+                                    var deserialize = utils.load_class.call(metadata[k].serialization)
+                                    that.serialize[k] = deserialize;
+                                    state[k] = deserialize.then((function(input) {
```
This is a little misleading.  When I first read it I didn't understand what was happening until I got down to the send sync msg method.  I suggest s/`deserialize`/`serializer` and s/`that.serialize`/`that.serializer`.  My thinking is that a `serializer` may contain both or either `serialize` and `deserialize` methods.

jdfreder commented on IPython/html/static/widgets/js/widget.js:
```diff
@@ -307,68 +350,70 @@ define(["widgets/js/manager",
+                    var value = state[key];
+                    if (value.buffer instanceof ArrayBuffer) {
+                        value = value.buffer;
+                    }
```
I don't think I understand the if statement above.  In case I do, is setting `.buffer` of a value a convention in Javascript?  If not, why would we want to support that?  It seems like value itself should be an ArrayBuffer if someone wants to use the binary msg spec.

jasongrout replied:
The DataView and various typed views have the ArrayBuffer as a .buffer attribute.  It's a convenience thing that you can give back a DataView or a typed array.  It's probably also fine to just insist that you give back an ArrayBuffer.  I'll look at this when I do some tests for the binary syncing.

jdfreder replied:
Thanks!  To be honest, I'm not familiar enough with either Types, I trust your judgement.

jdfreder commented on IPython/html/static/widgets/js/widget.js:
```diff
@@ -307,68 +350,70 @@ define(["widgets/js/manager",
+            })
+        },
+        
+        serialize: function(model, attrs) {
```
This name conflicts with the dict that you're using for the key->serializer promises dictionary (the one I think should be renamed to `serializers`).

jasongrout replied:
Great idea.  I made the change.

jdfreder commented on IPython/html/widgets/widget.py:
```diff
@@ -391,13 +417,13 @@ def _trait_to_json(self, x):
+            return [self._trait_to_json(v) for v in x], None
         elif isinstance(x, Widget):
-            return "IPY_MODEL_" + x.model_id
+            return "IPY_MODEL_" + x.model_id, None
```
As the next step, you should try using this new API to remove this and use custom (de)serializers to handle it.

jdfreder replied:
(next step- within this PR, not a separate PR)

jdfreder commented:
My biggest gripe about this design is that custom serializers are required for any variation of nested traits and I would argue that that means this doesn't scale well.  Additionally, as-is this supports `Unions` in one direction, from Python to Javascript, but you can't really say the same in reverse.

jasongrout commented:
The branch now minimally works for widget boxes (e.g., it correctly serializes and unserializes widget references).

I'll look at your comments tomorrow.  Thanks!

jasongrout commented:
@jdfreder - since any serializers/deserializers are supported, it would be easy to make default serializers that recurse into container traitlets, and thus experiment with different ways of doing things.

jdfreder commented:
@jasongrout I think this design is starting to shape up nicely.  I'll need to experiment with this branch by writing some custom serializers to see how it works for me.

jasongrout commented:
@jdfreder - I'm also writing some custom serializers to test things out.  Then I'll write some formal unit tests based on that.

jasongrout committed:
work-in-progress for custom js serializers

rgbkrk commented:
Definitely looking forward to this one, hope I can learn from it and use it soon.

jdfreder commented:
Wow, lots of changes since the last time I looked.  Everything is looking good so far.

jasongrout committed:
Delete the packing/unpacking tests, since we just use generic json now

jasongrout commented:
Still have some errors; I'll look at these tomorrow, though.

jasongrout commented:
I'm going to split this into a typo fix, the kernel.js fix, and then the widget stuff, so that it's easy to finish the widget stuff even after The Big Split.

minrk commented:
@jasongrout that sounds perfect, thanks.

jasongrout commented:
I split out the changes not dealing with widgets, and consolidated the changes dealing with widgets.

jasongrout committed:
Make message throttling test easier to debug

jasongrout committed:
Increment the pending_msgs counter immediately to avoid race conditions that send more messages

jasongrout committed:
Fix two calls to utils.reject (misplaced parens)

jasongrout commented:
@jdfreder - at a glance, it looks like the Travis failure is a Travis problem.  Is it easy to kick it?

jasongrout committed:
Simplify error handling for errors in sending sync messages from js

jdfreder commented:
> Is it easy to kick it?

Kicked :boom:

jasongrout committed:
Add the outline of a few new tests for the custom serialization

jasongrout committed:
Write tests for custom serialization

jasongrout committed:
Handle case when pixel values are undefined or null

jasongrout commented:
@jdfreder, @sylvaincorlay, I think this is basically ready for you guys to look at.  I'll be making minor changes (like making the quotes consistent in the tests), and fixing the missing parens in the test, but I think it's basically ready for comments.

jasongrout committed:
Make quoting consistent in the new tests

jdfreder commented on IPython/html/static/widgets/js/types.js:
```diff
@@ -0,0 +1,50 @@
+                } else {
+                    return Promise.resolve(value);
+                }
+            },
```
Curious, shouldn't there be a `serialize` function here for widgets?  If not for anything else for consistency?

jasongrout replied:
I don't think it's needed.  I changed the widgets to serialize themselves to their UUIDs, so the built-in parser will just work.  What I can do, though, is put in a comment.

jasongrout replied:
oh, wait; do you use `get_state` elsewhere?  You're right that I'm changing what get/set state are doing here, and pushing the serialization/deserialization work up.  Perhaps the serialization/deserialization should happen inside get/set_state still?

jdfreder commented on IPython/html/static/widgets/js/types.js:
```diff
@@ -0,0 +1,50 @@
+], function(utils){
+
+    return {
+        widget_serialization: {
```
Below you have `list_of_numbers`, maybe this should just be `widget`.

jasongrout replied:
I'm deleting list_of_numbers...

but I'll rename so the name is shorter (since it gets transferred across *every* *single* *time* a list of children is transferred.

jasongrout committed:
Correct documentation

jasongrout committed:
Convert some test code to python2/3 (add parens for print)

jasongrout commented on IPython/html/static/widgets/js/widget.js:
```diff
@@ -558,7 +611,7 @@ define(["widgets/js/manager",
              * Makes browser interpret a numerical string as a pixel value.
              */
-            if (/^\d+\.?(\d+)?$/.test(value.trim())) {
+            if (value && /^\d+\.?(\d+)?$/.test(value.trim())) {
```
@jdfreder - can you look at this?  I was getting test failures on occasion because `value` was `undefined` or `null` (sorry, forget which).  I didn't trace down the real problem yet, though.

jdfreder replied:
`null`, `undefined`, and `string` values all work fine, but a `number` value raises an `undefined` error.  I opened a PR against your branch that fixes this https://github.com/jasongrout/ipython/pull/8.

jasongrout commented:
For example's sake, here is another example custom serializer

```python
def _to_binary_list(x):
    import numpy
    return memoryview(numpy.array(x, dtype=float)), {'serialization': ('list_of_numbers', 'widgets/js/types')}

def _from_binary_list(x):
    import numpy
    a = numpy.frombuffer(x.tobytes(), dtype=float)
    return list(a)

list_of_numbers = {
    'from_json': _from_binary_list,
    'to_json': _to_binary_list
}
```

```javascript
list_of_numbers: {
    deserialize: function (value, model) {
        /* value is a DataView */
        /* create a float64 typed array */
        return new Float64Array(value.buffer)
    },
    serialize: function (value, model) {
        return value;
    },
}
```

jasongrout committed:
Delete unused custom serializer

jasongrout committed:
Delete unnecessary comment

jdfreder commented on IPython/html/static/widgets/js/widget.js:
```diff
@@ -162,30 +188,39 @@ define(["widgets/js/manager",
-            }
-            return state;
+            // Equivalent to Backbone.Model.toJSON()
+            return _.clone(this.attributes);
```
Doesn't this mean `get_state` will have widget model instances in it?  If it does, won't this break persistence?

jasongrout replied:
Do you mean just for attributes that have references to other widgets?  They should be serialized to their ids.

I'm checking manually to make sure widgets are serialized to their ids.  I'm puzzled that I don't see a toJSON() method implemented doing that (at one point, I had one, but I'm not sure where it has gone).

jasongrout replied:
(fixed widgets serializing to their ids in e0593c6.

jdfreder replied:
Thanks :)

jasongrout committed:
Fix serialization of models from js -> kernel

jdfreder commented on IPython/html/tests/widgets/widget.js:
```diff
@@ -113,20 +62,20 @@ casper.notebook_test(function () {
+        '        self.d = len(sync_data)',
+        'multiset = MultiSetWidget()',
+        'display(multiset)',
+        'print(multiset.model_id)'].join('\n'));
```
Nice trick, much better than \n everywhere!

jasongrout committed:
Change custom serialization to use custom models, rather than transmitting the serializer name across the wire

This separates the kernel and the js much more cleanly, and doesn't use as much space on the wire as well!

jasongrout committed:
Update widget test for new custom serialization changes

jasongrout commented:
@jdfreder: Okay, I'm much happier with the design now.  I think this is ready for review (again :).

jasongrout commented:
Yeah, it passed Travis!

minrk commented:
Looks very nice, @jasongrout. I'll let @jdfreder finish review and merge.

ellisonbg commented:
Pinging @jdfreder for another review pass...

jasongrout commented:
I haven't been able to run the binary websocket tests.  It would be great if someone could do so.

minrk commented:
@jasongrout binary websocket tests don't pass, but it's unrelated to this PR. It's just the API change from the switch to memoryviews. I opened #8200 to fix it.

jdfreder commented:
I'm reviewing this now

jdfreder commented on IPython/html/static/widgets/js/init.js:
```diff
@@ -14,18 +14,33 @@ define([
+            if (module.hasOwnProperty(target_name)) {
+                if (endsWith(target_name, "View")) {
+                    widgetmanager.WidgetManager.register_widget_view(target_name, module[target_name]);
+                } else if (endsWith(target_name, "Model")) {
```
I don't think we should be assuming, even within our own project, that the class name's suffix correlates to the parent class.  Here's something I whipped up that can be used to check for inheritance of backbone classes:

```js
require(['backbone', 'widgets/js/widget', 'widgets/js/widget_button'], function(backbone, widget, button) {
    var isinstance = function(child, parent) {
        child = child.__super__;
        while (child !== null) {
            if (child === parent.prototype) return true;
            child = child.__proto__;
        }
        return false;
    };

    console.log(isinstance(button['ButtonView'], widget['DOMWidgetView']));
    console.log(isinstance(button['ButtonView'], widget['WidgetView']));
    console.log(isinstance(button['ButtonView'], widget['WidgetModel']));
});
```

true, true, false

jasongrout replied:
class inheritance is a much better idea than working on the name.  Does Backbone not have something like isinstance?

jdfreder replied:
I don't know, TBH I didn't look :(

jdfreder replied:
Looks like instanceof works- I haven't tested it yet though.  http://stackoverflow.com/a/11125117

jdfreder replied:
@jasongrout yup, much simpler:

```js
button['ButtonView'].prototype instanceof widget['DOMWidgetView']
//true
button['ButtonView'].prototype instanceof widget['WidgetView']
//true
button['ButtonView'].prototype instanceof widget['WidgetModel']
//false
```

jdfreder commented on IPython/html/static/widgets/js/widget_box.js:
```diff
@@ -4,10 +4,41 @@
-], function(widget, $, utils){
+], function(widget, $, _, utils){
     "use strict";
+    var unpack_models = function unpack_models(value, model) {
```
You should probably export this function so custom widget authors don't have to rewrite this function.

jasongrout replied:
Good point!

jasongrout replied:
Done.

jdfreder commented:
@jasongrout overall I'm happy with this design too.  If later we find that we need something more dynamic, like box typing that @sylvaincorlay suggested, we'll cross that bridge when/if we come to it.

Just two comments inline.

Commit:
Register widget models and views that have the right inheritance, rather than the right name.

Thanks to @jdfreder for this suggestion.

Commit:
Export the widget unpack_models function for other custom classes to use

Thanks to @jdfreder for this suggestion.

jasongrout commented:
@jdfreder: I think I addressed your two points.  Thanks!  I'll wait to make sure the tests pass as well.

jasongrout commented:
@jdfreder: Looks like tests passed....

jasongrout commented:
This does change the number of arguments in custom message callbacks; I'm not sure where to note that.

jdfreder commented:
Awesome!  Go ahead and add a what's new rst in https://github.com/ipython/ipython/tree/master/docs/source/whatsnew/pr then I think this is :+1:

rgbkrk commented:
Woohoo!

jdfreder commented:
@jasongrout IIRC you're on vacation right?  Seems like we have an even greater timezone difference today!  I'm going to merge this and write the what's new rst for you.

SylvainCorlay commented:
@jdfreder @jasongrout it seems that it is breaking the loading of custom models. I am looking into this now.

jasongrout commented:
@SylvainCorlay, thanks.  @jdfreder, thanks for moving ahead without waiting for me.

One incompatible change is that the custom message handlers in python now take a new argument, `buffers`.  See the change for the click handler for buttons, for example.

jasongrout commented:
@SylvainCorlay, also note that models which have fields that encode references to other models need to now explicitly declare the serializer for that field (like in widget_box's children attribute).  This, for example, will necessitate some changes to pythreejs.

SylvainCorlay commented:
@jasongrout thanks for the pointers. 
This actually forces all custom widgets holding references to other models to have a custom model.

jasongrout commented:
Right.  I think that is much better than the hack we had in serialization before.

SylvainCorlay commented:
I agree. Do you think that we still need the `IPY_MODEL_` prefix?

jasongrout commented:
Probably less of a need.  We could shorten it, but I don't think we should get rid of it.

jasongrout commented:
@tacaswell, I forgot to mention to you that this PR enables binary buffers in IPython comm messages, which may be useful to the nbagg backend.

tacaswell commented:
attn @pelson 

That does seem handy.  This means we can send the png as binary rather than as a base64 string?

jasongrout commented:
Exactly
